Utilities - Add NodeNetworkConfigurationPolicy resource to libs/net

Anatw opened the pull request:
##### Short description:
Introducing a new NodeNetworkConfigurationPolicy (NNCP) resource to the libs/net library. This resource is meant to replace usage oc NNCP in the wrapper, which has large amount of functions added to it, which makes it difficult to add new helpers and functionalities.

##### More details:
The new NNCP is used in an existing test - test_linux_bridge_functionality.

##### What this PR does / why we need it:
As part of the initiative to create simpler resources and reduce dependency, this PR introduces a new NNCP resource in the lib directory.

cnv-qe-bot1 commented:
Report bugs in [Issues](https://github.com/myakove/github-webhook-server/issues)

The following are automatically added:
 * Add reviewers from OWNER file (in the root of the repository) under reviewers section.
 * Set PR size label.
 * New issue is created for the PR. (Closed when PR is merged/closed)
 * Run [pre-commit](https://pre-commit.ci/) if `.pre-commit-config.yaml` exists in the repo.

Available user actions:
 * To mark PR as WIP comment `/wip` to the PR, To remove it from the PR comment `/wip cancel` to the PR.
 * To block merging of PR comment `/hold`, To un-block merging of PR comment `/hold cancel`.
 * To mark PR as verified comment `/verified` to the PR, to un-verify comment `/verified cancel` to the PR.
        verified label removed on each new commit push.
 * To cherry pick a merged PR comment `/cherry-pick <target branch to cherry-pick to>` in the PR.
    * Multiple target branches can be cherry-picked, separated by spaces. (`/cherry-pick branch1 branch2`)
    * Cherry-pick will be started when PR is merged
 * To build and push container image command `/build-and-push-container` in the PR (tag will be the PR number).
    * You can add extra args to the Podman build command
        * Example: `/build-and-push-container --build-arg OPENSHIFT_PYTHON_WRAPPER_COMMIT=<commit_hash>`
 * To add a label by comment use `/<label name>`, to remove, use `/<label name> cancel`
 * To assign reviewers based on OWNERS file use `/assign-reviewers`
 * To check if PR can be merged use `/check-can-merge`
 * to assign reviewer to PR use `/assign-reviewer @<reviewer>`

<details>
<summary>Supported /retest check runs</summary>

 * `/retest tox`: Retest tox
 * `/retest build-container`: Retest build-container
 * `/retest all`: Retest all

</details>

<details>
<summary>Supported labels</summary>

 * hold
 * verified
 * wip
 * lgtm

</details>

polarion-jenkins reviewed (approved):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/6320

polarion-jenkins reviewed (dismissed):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/6324

polarion-jenkins reviewed (dismissed):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/6326

yossisegev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,250 @@
+        return self.res["spec"]["desiredState"]["interfaces"]
+
+
+def replace_underscore_with_hyphen_and_filter_none_values(data: list[tuple[str, Any]]) -> dict[str, Any]:
```
2 things:
1. The name of this function is confusing. It implies that it only performs some cosmetic actions, but it actually produces a dict out of a list object.
2. The part of replacing underscores seems like the function is covering-up for the callers misbehavior. The caller should create a proper NNCP object, which also includes creating dict keys according to the kubernetes rules.

Anatw replied:
1. The actual conversion is known from the caller (in to_dict), but the function doesn't care who calls it and should declare itself correctly. Ack, I'll rename.
2. Not exactly, the caller uses dataclass, which behave in a certain non-kuberneted way -> the field names (keys) can only include underscores and not hyphens. Later we need to alter it according to our kubernetes needs.

yossisegev replied:
2. Ack, thank you for explaining.

Anatw commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,250 @@
+    class Conditions:
+        class Type:
+            DEGRADED = "Degraded"
+            AVAILABLE = "Available"
```
I didn't use it currently, where do you think it is needed?

Anatw commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,250 @@
+        self._desired_state = desired_state
+        self._node_selector = node_selector
+
+    def deploy(self, wait: bool = False) -> "NodeNetworkConfigurationPolicy":
```
It was added due to an error I got but now that I remove it - there are no complains. I ran it without the quotation marks and it runs fine. Thanks.

EdDev replied:
> It was added due to an error I got but now that I remove it - there are no complains. I ran it without the quotation marks and it runs fine. Thanks.

I think the question was about correctness and not about linter warnings.
IMO the fact the linter was quiet is not that convincing it is correct.

---

@yossisegev , based on [PEP-0484](https://peps.python.org/pep-0484/#forward-references):
> When a type hint contains names that have not been defined yet, that definition may be expressed as a string literal, to be resolved later.

So this is how it is done for these cases.
Once [PEP-794](https://peps.python.org/pep-0749/) lands, you'll be able and define the same without the quotes.

Anatw commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,250 @@
+        if errors:
+            err_msg += f"{nnce_prefix}: {errors[0]}"
+            if len(errors) > 1:
+                err_msg += errors[-1]
```
Ack. I'll log all the errors (As well as libnmstate_err 2 lines bellow).

polarion-jenkins reviewed (approved):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/6338

Anatw commented:
[Change](https://github.com/RedHatQE/openshift-virtualization-tests/compare/e405d3e6e61fde236987317ee36299b99d49b7c5..a9449d527af9702eda18fcf8ffed99596cf2d496): refactor code according to CR:
 - When encounter NNCE error - log all errors and not just the first and last.
 - Rename  list to dict serialization function.
 - Remove parentheses from type return value.
 - Refactor NodeNetworkConfigurationPolicy docstring.

polarion-jenkins reviewed (dismissed):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/6340

Anatw commented:
[Change](https://github.com/RedHatQE/openshift-virtualization-tests/compare/0b5a42fa5c94d2c6fe8241b10a59c492744a6a62..49a0112a061c27ea7fedcbcdde0f366b290ba463): fix some pre-commit issues.

polarion-jenkins reviewed (dismissed):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/6342

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,246 @@
+
+        raise NNCPConfigurationFailed(f"{self.name}. None of the conditions were met.")
+
+    def _absent_interface(self) -> None:
+        desired_state = {}
+        if self._desired_state.interfaces:
+            for _iface in self.res["spec"]["desiredState"]["interfaces"]:
+                _iface["state"] = Resource.Interface.State.ABSENT
+
+            desired_state["interfaces"] = self.res["spec"]["desiredState"]["interfaces"]
+
+        ResourceEditor(patches={self: {"spec": {"desiredState": desired_state}}}).update()
```
1. "absent" is not a verb. Perhaps "remove" or "delete": `_delete_interfaces`.
2. You should use the new data structure, not the raw dict. When you patch, then convert it to a dict.
3. Use dict comprehension to build the new data.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,246 @@
+
+        ResourceEditor(patches={self: {"spec": {"desiredState": desired_state}}}).update()
+
+    def clean_up(self, wait: bool = True, timeout: int | None = None) -> bool:
```
The implementation below does not look correct to me.
The non-used arguments do not help as well.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,246 @@
+        return self.res["spec"]["desiredState"]["interfaces"]
+
+
+def list_to_dict_serialize(data: list[tuple[str, Any]]) -> dict[str, Any]:
```
"list to dict"? Say what?
Maybe call it `keyval_normalization` and document what it does (filter out none values and converts key characters from `_` to `-`).

EdDev commented on tests/network/cluster_addons_operator/test_network_addons_operator.py:
```diff
@@ -16,6 +17,16 @@
+    DesiredState,
+    Interfaces,
+    NodeNetworkConfigurationPolicy,
+    _default_desired_ipv4,
+    _default_desired_ipv6,
```
?!

EdDev reviewed (changes requested):
I have not covered everything.
Try to treat the points raised until now and I'll revisit to continue then.

dbasunag reviewed (commented):
Why would we want to duplicate wrapper code to the test repo? I am not sure I understand what you are trying to do.

dbasunag commented:
/hold

copying over code from wrapper is not productive. If you don't like how the code is organized there, please create a PR there and work with the wrapper mantainers. I won't take this changes here.

Anatw commented:
[Changes](https://github.com/RedHatQE/openshift-virtualization-tests/compare/49a0112a061c27ea7fedcbcdde0f366b290ba463..779cf1f18a8c03e59edc922baed19b97b3871664):
 - Change the order of dataclasses for more readability.
 - Have a base IP class for use in IPv4 and IPv6.
 - In to_dict, if no node_selector was provided - set it to be an empty list.
 - In DesiredState, set interfaces to None instead of an empty list - the caller should specify that he want's an empty list.
 - Remove NNCE usage for error reason - must gather should take care of that, we only need to print the status 
    (success/failure).
 - In wait_for_status_success:
         - replace usage of TimeoutSampler with retry decorator.
         - Refine the logic  and only print the status (success/failure).
 - Rename _absent_interface to _delete_interfaces and work with the _desired_state instead of the dict.
 - Remove unused parameters from the clean_up signature.
 - Rename list_to_dict_serialize to dict_normalization and add a short description.
 - Remove redundant functions - _default_desired_ipv4 and _default_desired_ipv6.

polarion-jenkins reviewed (approved):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/6815

Anatw commented:
> Why would we want to duplicate wrapper code to the test repo? I am not sure I understand what you are trying to do.
> 
> /hold
> 
> copying over code from wrapper is not productive. If you don't like how the code is organized there, please create a PR there and work with the wrapper mantainers. I won't take this changes here.

Agreed, my purposes weren't clear enough, sorry for that.
Now the duplication is gone, and this new NNCP is more plain, focused on what we need starting from the base, what will allow us to more easily build upon it. And example for this you'll be able to see in my upcoming PR, adding localnet support to this NNCP, what we found really hard to do with the existing NNCP resource in the wrapper.
I'd appreciate if you can review it again and remove the hold if you agree.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,246 @@
+
+
+@dataclass
+class Port:
```
+1

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,179 @@
+    def _delete_interfaces(self) -> None:
+        desired_state = {}
+        if self._desired_state.interfaces:
+            for _iface in self._desired_state.interfaces:
```
Since _iface is a local variable within a private method, it doesn't need to be prefixed with an underscore.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,179 @@
+                nncp_condition["reason"] == self.Conditions.Reason.SUCCESSFULLY_CONFIGURED
+                and nncp_condition["status"] == self.Condition.Status.TRUE
+            ):
+                self.logger.info(f"NNCP {self.name} configured Successfully")
```
```suggestion
                self.logger.info(f"{self.kind}/{self.name} configured successfully")
```

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,179 @@
+                return nncp_condition
+
+            elif (
+                nncp_condition["reason"] == no_match_node_condition_reason
```
We need to check that the status is True in addition to the reason to accurately assess the condition's state.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,179 @@
+                or nncp_condition["reason"] == failed_condition_reason
+            ):
+                raise NNCPConfigurationFailed(
+                    f"NNCP {self.name} failed to configure with error: {nncp_condition['reason']}\n"
```
ditto
Plus, does the '\n' really required?
You catch the exception at:
https://github.com/RedHatQE/openshift-virtualization-tests/blob/779cf1f18a8c03e59edc922baed19b97b3871664/libs/net/nodenetworkconfigurationpolicy.py#L168
Then you use the logger library, which handle it by default.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,179 @@
+
+    def _delete_interfaces(self) -> None:
+        desired_state = {}
+        if self._desired_state.interfaces:
```
The if self._desired_state.interfaces check is redundant since interfaces is always a list (not optional) and an empty list is handled by the for loop. It can be removed for cleaner code.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,179 @@
+        return self.delete()
+
+    def interfaces(self) -> list[dict]:
+        return self.res["spec"]["desiredState"]["interfaces"]
```
Add a check to ensure "interfaces" exists in self.res["spec"]["desiredState"] before accessing it to prevent potential KeyError.
For example, in localnet NNCP, we do not set interfaces.
desiredState is required anyway by the API, so no need to check it.

Anatw commented:
[Change](https://github.com/RedHatQE/openshift-virtualization-tests/compare/779cf1f18a8c03e59edc922baed19b97b3871664..8e3c3e4e07e7e2449cfadb3d0c6239fa56497edc):
 - Change the order of dataclasses and methods for better readability.
 - Remove leading underscore in _delete_interfaces.
 - Add try/except in interfaces
 - Change logs with "NNCP" to use self.kind.
 - In wait_for_status_success, check that the status is True before going over the conditions.

polarion-jenkins reviewed (approved):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/6878

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+
+
+@dataclass
+class Interfaces:
```
I think the name should be singular.

sbahar619 replied:
Just to be clear, the suggestion of @EdDev as I understand is:
```suggestion
class Interface:
```

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+                    self.logger.info(f"{self.kind} {self.name} configured successfully")
+                    return nncp_condition
+
+                elif (
```
Because the prev body returns, this one can be a simpler `if`.

This could be converted to a `matcher`, but not a must.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+        sleep=TIMEOUT_5SEC,
+    )
+    def wait_for_status_success(self) -> list[dict]:
+        failed_condition_reason = self.Conditions.Reason.FAILED_TO_CONFIGURE
+        no_match_node_condition_reason = self.Conditions.Reason.NO_MATCHING_NODE
+        nncp_conditions = self.instance.status.conditions
```
What is the purpose of these (3) variables?

sbahar619 replied:
Identifying known exceptions helps reduce the method's wait time, so I think you should keep that approach. As for the three variables, they’re only used once, so there's no strong reason to define them separately. From a readability perspective, the class-based references are already clear in context.

Anatw replied:
I'll remove them, they are used twice but it doesn't really matter and I agree it looks messyer with the added variables.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+
+    def interfaces(self) -> list[dict]:
+        try:
+            return self.res["spec"]["desiredState"]["interfaces"]
```
You are reading this from the raw cash. Is this what you intended or you are expecting to see what is on the NNCP object now? Can you please clarify?

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+        except Exception as exp:
+            self.logger.error(exp)
+
+        return self.delete()
```
Why it is not deleted through the context manager? Why would we want to remove the object through special methods?

EdDev replied:
Related to prev thread: https://github.com/RedHatQE/openshift-virtualization-tests/pull/218/files#r1954511796

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+
+        ResourceEditor(patches={self: {"spec": {"desiredState": desired_state}}}).update()
+
+    def clean_up(self) -> bool:
```
Is something using this one? I could not detect its usage.

EdDev replied:
I guess you mean it is called by a parent class?
If that is the case, why you are not first doing the interface deleting and then calling the parent method?
I fail to understand why it needs to be re-implemented here.

Anatw replied:
I mean that this function is called as part of the cleanup, and in this step (the cleanup) we need 2 things to happen:
1. Interfaces are set as absent and the NNCP is configured successfully.
2. The NNCP object is deleted.
Step 1 is the reason I need to overwrite the parent tear-down class.
I can change return self.delete() to return super().clean_up(), if this is what you mean.

EdDev replied:
You can resolve this one, I see you called the parent method.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,179 @@
+        self,
+        name: str,
+        desired_state: DesiredState,
+        client: DynamicClient | None = None,
```
Currently the client is not used anyway, so we can remove it and add it if some caller in future will request it.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+    def wait_for_status_success(self) -> list[dict]:
+        failed_condition_reason = self.Conditions.Reason.FAILED_TO_CONFIGURE
+        no_match_node_condition_reason = self.Conditions.Reason.NO_MATCHING_NODE
+        nncp_conditions = self.instance.status.conditions
```
Since this code is inside a method of the NodeNetworkConfigurationPolicy class, the nncp_ prefix in nncp_conditions seems redundant. The context already makes it clear that these conditions belong to an NNCP. Consider renaming it to just conditions for better readability.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+        failed_condition_reason = self.Conditions.Reason.FAILED_TO_CONFIGURE
+        no_match_node_condition_reason = self.Conditions.Reason.NO_MATCHING_NODE
+        nncp_conditions = self.instance.status.conditions
+        for nncp_condition in nncp_conditions:
```
ditto

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+        no_match_node_condition_reason = self.Conditions.Reason.NO_MATCHING_NODE
+        nncp_conditions = self.instance.status.conditions
+        for nncp_condition in nncp_conditions:
+            if nncp_condition["status"] == f"{True}":
```
Why not using:
```suggestion
            if nncp_condition["status"] == self.Condition.Status.TRUE:
```

To be more consistent like you used here:
https://github.com/RedHatQE/openshift-virtualization-tests/blob/8e3c3e4e07e7e2449cfadb3d0c6239fa56497edc/libs/net/nodenetworkconfigurationpolicy.py#L122

sbahar619 replied:
I suggest filtering the conditions first based on this logic to eliminate the need for a nested if block:
```
conditions = [
        condition for condition in self.instance.status.conditions if condition["status"] == self.Condition.Status.TRUE
    ]
```

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+        try:
+            return self.res["spec"]["desiredState"]["interfaces"]
+        except KeyError:
+            return []
```
I suggest adding a log warning with the exception message so it's clear why an empty list is returned.

EdDev replied:
I think it is clear and a log is surely not a must for a an acceptable scenario.

If it is not clear, the scenario is that the definition has not defined any chances of interfaces, but something else. It is valid not to define the interfaces AFAIK.

sbahar619 replied:
ack.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+            if nncp_condition["status"] == f"{True}":
+                if (
+                    nncp_condition["reason"] == self.Conditions.Reason.SUCCESSFULLY_CONFIGURED
+                    and nncp_condition["status"] == self.Condition.Status.TRUE
```
This condition check is redundant, you already check for it in the if statement above.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+                    or nncp_condition["reason"] == failed_condition_reason
+                ):
+                    raise NNCPConfigurationFailed(
+                        f"{self.kind} {self.name} failed to configure with error: {nncp_condition['reason']}\n"
```
Please log the entire condition dictionary before raising the exception. This will expose additional information, such as condition.message, making debugging easier.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+            AVAILABLE = "Available"
+
+        class Reason:
+            CONFIGURATION_PROGRESSING = "ConfigurationProgressing"
```
Not used, consider removing it.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+    api_group = Resource.ApiGroup.NMSTATE_IO
+
+    class Conditions:
+        class Type:
```
This entire class is not used.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+
+        return self.delete()
+
+    def interfaces(self) -> list[dict]:
```
The interfaces method is only used in net_add_op_br1test_nad to get the interface name. Instead, define a constant to improve clarity, avoid unnecessary calls, and prevent index errors. Since NNCP can have multiple interfaces, setting it explicitly is safer.

EdDev replied:
I do not understand this comment.
Try to give an example of how you want it to be used... show the current usage and the suggested one.

sbahar619 replied:
I suggest to define constant for the interface name, something like:
```
BRIDGE_NAME = "net-add-br"
```
Then use it directly and remove the interfaces method.
We already know what is the interface name, I don't see reason to fetch it dynamically.

EdDev replied:
Ohh my... all this wording is basically: This method is currently redundant, please see comment at usage.

EdDev replied:
It may be correct, but it is hard to reason for it now.
Introduce it when it is needed so we can examine if it makes sense or not based on the usage.

Anatw replied:
Ack, I'll remove it.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+            return []
+
+
+def dict_normalization(data: list[tuple[str, Any]]) -> dict[str, Any]:
```
The dict_normalization function is only used in to_dict() and is defined outside the class scope. Consider replacing it with a lambda function inside asdict() to simplify the code and keep the logic contained within the method.

EdDev replied:
I disagree.
The function is useful for readability to understand what it does. It has a formal doctext to explain itself as well. Inlining it, will be cryptic.

If you insist, you can make it a `staticmethod`, something the maintainers like more.

sbahar619 replied:
ack, staticmethod is fine.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+
+    def _delete_interfaces(self) -> None:
+        desired_state = {}
+        if self._desired_state.interfaces:
```
The check if self._desired_state.interfaces: is redundant since interfaces is always a list (even if empty). The method can be simplified by removing this condition and directly iterating over self._desired_state.interfaces.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+    def _delete_interfaces(self) -> None:
+        desired_state = {}
+        if self._desired_state.interfaces:
+            for iface in self._desired_state.interfaces:
```
I suggest to refactored the _delete_interfaces() method to avoid two loops by combining the state update and dictionary conversion into a single loop. This improves efficiency and keeps the code cleaner.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,183 @@
+
+            desired_state["interfaces"] = [asdict(iface) for iface in self._desired_state.interfaces]
+
+        ResourceEditor(patches={self: {"spec": {"desiredState": desired_state}}}).update()
```
Currently, the patch will always be executed, regardless of whether self._desired_state.interfaces is empty or not, since the patch logic is outside any conditional control. This could result in unnecessary updates when no interfaces are present.

polarion-jenkins reviewed (dismissed):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/6961

Anatw commented:
[Change](https://github.com/RedHatQE/openshift-virtualization-tests/compare/8e3c3e4e07e7e2449cfadb3d0c6239fa56497edc..f4ea655b19305175a64ffad900570fae547d7443):
 - Replace generic timeout with specific one.
 - Rename Interfaces to Interface.
 - Remove class Type.
 - Removed CONFIGURATION_PROGRESSING Reason.
 - Remove client parameter from init.
 - Renamed some inner function variables.
 - In wait_for_status_success:
         - Use self.Condition.Status.TRUE.
         -  remove redundant  Status check.
         - Replace elif with if.
        - Raise the entire nncp_condition instead of just the reason.
 - In _delete_interfaces:
         - Remove redundant if self._desired_state.interfaces.
         - Convert 2 loops into one.
         - Run ResourceEditor only when needed.
 - In interfaces, check the live resource instead of res.
 - Refactor dict_normalization into a staticmethod.

Anatw commented:
[Change](https://github.com/RedHatQE/openshift-virtualization-tests/compare/8e3c3e4e07e7e2449cfadb3d0c6239fa56497edc..f4ea655b19305175a64ffad900570fae547d7443): Remove #.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,179 @@
+@dataclass
+class IP:
+    auto_dns: bool = True
+    dhcp: bool = False
```
You should disable DHCP when creating an instance instead of making it the default in the class.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,171 @@
+            FAILED_TO_CONFIGURE = "FailedToConfigure"
+            NO_MATCHING_NODE = "NoMatchingNode"
+
+    def __init__(
```
Please add **kwargs as an argument to expose all inherited attributes, allowing the caller to set them. Make sure to pass it using super().__init__(**kwargs).

sbahar619 replied:
Interesting, I don't mind continue without it, but can you please give an example of how it can break API?

EdDev replied:
1. It is a "no limit" API, i.e. anything is accepted. When anything is accepted, anything can be ignored, blocked or treated differently. There is no ability to track it well, because it depends on how deep the inheritance is or how deep this is used in the object.
2. There is no proper API documentation.
3. It is impossible to provide recommendation to the users on how to use it. When the arguments are explicit, it not only comes with documentation, but it also comes with how the author intended it to be used.
4. In most cases I saw, it breaks the Interface Segregation Principle [1].

[1] https://en.wikipedia.org/wiki/Interface_segregation_principle

sbahar619 replied:
Thanks!
@EdDev 
Noted.

sbahar619 replied:
I don't have this option to resolve the thread.

EdDev replied:
Only the author has it (crazy GitHub thing), so you can just express you are satisfy with a message or thumb-up and then the author can resolve it.

sbahar619 replied:
ack, this thread can be resolved.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,171 @@
+                        f"{self.kind} {self.name} failed to configure on condition: {condition}\n"
+                    )
+
+        raise NNCPConfigurationFailed(f"{self.name}. None of the conditions were met.")
```
I don't think raising this exception is correct. If none of the conditions were met, it doesn't necessarily mean the NNCP failed to configure—it could still be in progress.

Anatw replied:
We will get here if we reached timeout - 90 seconds. In this stage it shouldn't be in progress.

sbahar619 replied:
No, it will get here after it will finish to iterate over all the conditions in self.instance.status.conditions.

Anatw commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,171 @@
+    interfaces: list[Interface]
+
+
+class NodeNetworkConfigurationPolicy(Resource):
```
Ack, thank you.

Anatw commented:
[Changes](https://github.com/RedHatQE/openshift-virtualization-tests/compare/9a7d2a2d9853bc094ea7732a6e60208c5153eec7..720462442bf935db1e4bcbfe06b502537c03ef1a):
 - Rename NodeNetworkConfigurationPolicy in net/libs to NetLibsNodeNetworkConfigurationPolicy.
 - NetLibsNodeNetworkConfigurationPolicy Now inherits from the NodeNetworkConfigurationPolicy in ocp_resources.node_network_configuration_policy_latest, instead of from Resource.
 - In wait_for_status_success, remove 3 variables (failed_condition_reason, no_match_node_condition_reason, nncp_conditions) and use their content directly.
 - Remove interfaces method and use it's content directly in net_add_op_br1test_nad.
 - In the clean_up method replace return self.delete() with return super().clean_up().

polarion-jenkins reviewed (approved):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/7146

Anatw commented:
I forgot to push the changes on main independently, sorry.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,163 @@
+        self._desired_state = desired_state
+        self._node_selector = node_selector
+
+    @retry(
```
The wait_for_status_success function is not really waiting, because if in a case the expected status is not present, the function will raise NNCPConfigurationFailed which will stop the waiting functionality of TimeoutSampler.
You need to specify the group of exception that should be ignored and let the function managed by TimeoutSampler to continue using exceptions_dict in the decorator.

polarion-jenkins reviewed (dismissed):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/7263

Anatw commented:
[Changes](https://github.com/RedHatQE/openshift-virtualization-tests/compare/d51e2442d3b37341745ee3e515af47af70637a1d..a46f3067feecf4d746971031cfdf5a48fdfd0b36):
 - Add link to API.
In nodenetworkconfigurationpolicy:
 - Change in the IP class, dhcp has no default value.
 - Add a new costume exception - NNCPSuccessStatusNotMet
In wait_for_status_success:
 - In the retry decorator - add an exceptions_dict.
 - Instead of raising NNCPConfigurationFailed, raise a more specific raise - NNCPSuccessStatusNotMet (that was also added to the exceptions_dict).
In test_network_addons_operator:
 - In net_add_op_bridge_device, send dhcp=False in ipv4 and ipv6.
 - In net_add_op_br1test_nad, fix interface_name call.

polarion-jenkins reviewed (approved):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/7278

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,167 @@
+    pass
+
+
+class NetLibsNodeNetworkConfigurationPolicy(NodeNetworkConfigurationPolicy):
```
This class doesn't introduce any additional functionality beyond NodeNetworkConfigurationPolicy. We can use the base class directly instead of maintaining a redundant subclass. Should we consider removing it?

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,167 @@
+        Args:
+            name (str): Name of the NodeNetworkConfigurationPolicy object.
+            desired_state (DesiredState): Desired policy configuration - interface creation, modification or removal.
+            client (DynamicClient, optional): Optional DynamicClient to use.
```
Leftover.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,167 @@
+    @retry(
+        wait_timeout=WAIT_FOR_STATUS_TIMEOUT, sleep=WAIT_FOR_STATUS_SLEEP, exceptions_dict={NNCPSuccessStatusNotMet: []}
+    )
+    def wait_for_status_success(self) -> list[dict]:
+        for condition in self.instance.status.conditions:
+            if condition["status"] == self.Condition.Status.TRUE:
+                if condition["reason"] == self.Conditions.Reason.SUCCESSFULLY_CONFIGURED:
+                    self.logger.info(f"{self.kind} {self.name} configured successfully")
+                    return condition
+
+                if (
+                    condition["reason"] == self.Conditions.Reason.NO_MATCHING_NODE
+                    or condition["reason"] == self.Conditions.Reason.FAILED_TO_CONFIGURE
+                ):
+                    raise NNCPConfigurationFailed(
+                        f"{self.kind} {self.name} failed to configure on condition: {condition}\n"
+                    )
+
+        raise NNCPSuccessStatusNotMet(f"{self.name}. None of the conditions were met.")
```
Suggestion:
Refactored to reduce nesting by filtering conditions upfront, improving readability and maintainability. This makes the logic clearer and avoids redundant checks.

```suggestion
    def wait_for_status_success(self) -> dict:
        conditions = [
            condition
            for condition in self.instance.status.conditions
            if condition["status"] == self.Condition.Status.TRUE
        ]
    
        for condition in conditions:
            if condition["reason"] == self.Conditions.Reason.SUCCESSFULLY_CONFIGURED:
                self.logger.info(f"{self.kind} {self.name} configured successfully")
                return condition
    
            if condition["reason"] in {
                self.Conditions.Reason.NO_MATCHING_NODE,
                self.Conditions.Reason.FAILED_TO_CONFIGURE,
            }:
                raise NNCPConfigurationFailed(
                    f"{self.kind} {self.name} failed to configure on condition: {condition}\n"
                )
    
        raise NNCPSuccessStatusNotMet(f"{self.name}. None of the conditions were met.")
```

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,167 @@
+
+        raise NNCPSuccessStatusNotMet(f"{self.name}. None of the conditions were met.")
+
+    def deploy(self, wait: bool = False) -> "NodeNetworkConfigurationPolicy":
```
I recommend not redefining the deploy method, as it complicates the context manager. The context manager should focus on deploying the resource, not waiting for the expected status. If we need to call __exit__ ourselves, we've likely done something wrong, as the context manager should handle that automatically. Instead, we can let the user call the wait logic separately if needed.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,167 @@
+        desired_state["interfaces"] = [
+            asdict(iface)
+            for iface in self._desired_state.interfaces
+            if setattr(iface, "state", Resource.Interface.State.ABSENT) is None
```
```suggestion
            if setattr(iface, "state", self.Interface.State.ABSENT) is None
```

sbahar619 replied:
The is None check after setattr is redundant since setattr always returns None. Could you clarify its purpose? It can likely be removed for simplicity.

sbahar619 commented on tests/network/cluster_addons_operator/test_network_addons_operator.py:
```diff
@@ -149,21 +160,33 @@ def check_components(network_addons_config_scope_session):
+        interfaces=[
+            Interface(
+                name="net-add-br",
+                state=Resource.Interface.State.UP,
```
```suggestion
                state=NetLibsNodeNetworkConfigurationPolicy.Interface.State.UP,
```
Then you can also remove the import of Resource.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,167 @@
+        if desired_state["interfaces"]:
+            ResourceEditor(patches={self: {"spec": {"desiredState": desired_state}}}).update()
+
+    def clean_up(self) -> bool:
```
The same comment that I add for deploy:
I don't think we need to handle it with context manager, let the user handle it.
In the test fixture you can use the wait for success at setup, and delete interfaces for teardown.

EdDev replied:
The NNCP cleanup must make sure all the parts that have been defined by it are teardown by it.
Trusting that the caller will do it is unsafe.

Anatw commented:
Change: rebase on main.

polarion-jenkins reviewed (approved):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/7384

polarion-jenkins commented:
D/S test `tox -e verify-tc-requirement-polarion` failed: cnv-tests-tox-executor/7383

Anatw commented:
[Changes](https://github.com/RedHatQE/openshift-virtualization-tests/compare/c57245499f4656d4e7a87b120e3e644c6fbffdee..a4b02b7db0fd8351620a22bb6541f6ec2c744a37):
 - Remove redundant parameter in the docstring of NetLibsNodeNetworkConfigurationPolicy.
 - Refactor wait_for_status_success first for and if into a separate conditions list comprehension.
 - Remove the deploy method. Since we don't need to wait for status success as part of the deploy, but this should be waited by the user if needed, so there is no need to write over the super deploy.
 - In _delete_interfaces, refactor the single list comprehension into 2 loops, to avoid the use of "is None" that is failing pre-commit.
 - In net_add_op_bridge_device, replace usage of Resource to usage of NetLibsNodeNetworkConfigurationPolicy
 - In net_add_op_bridge_device, add a call to wait_for_status_success, which is now removed from the deploy method and is part of the user responsibility.

polarion-jenkins reviewed (approved):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/7388

polarion-jenkins commented:
D/S test `tox -e verify-tc-requirement-polarion` failed: cnv-tests-tox-executor/7387

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,159 @@
+# API: https://nmstate.io/devel/yaml_api.html#dhcp
```
Why you reference the API here?
Why dhcp specifically?
You should reference the whole API in the right place:
DesiredState

sbahar619 replied:
This nmstate API specifications is only relevant for desiredState API, so you also need to move it there.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,159 @@
+    NodeNetworkConfigurationPolicy object.
+    """
+
+    api_group = Resource.ApiGroup.NMSTATE_IO
```
Redundant, already defined by base class.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,159 @@
+
+
+@dataclass
+class STP:
```
Please arrange all classes here from top to bottom per their reference.
At the top, place `DesiredState` and from there per their order.

I will like to read this from top as a story.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,159 @@
+        wait_timeout=WAIT_FOR_STATUS_TIMEOUT, sleep=WAIT_FOR_STATUS_SLEEP, exceptions_dict={NNCPSuccessStatusNotMet: []}
+    )
+    def wait_for_status_success(self) -> list[dict]:
+        conditions = [
+            condition
+            for condition in self.instance.status.conditions
+            if condition["status"] == self.Condition.Status.TRUE
+        ]
```
This can be a generator.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,159 @@
+                or condition["reason"] == self.Conditions.Reason.FAILED_TO_CONFIGURE
+            ):
+                raise NNCPConfigurationFailed(
+                    f"{self.kind} {self.name} failed to configure on condition: {condition}\n"
```
Why newline in the end?

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,159 @@
+                    f"{self.kind} {self.name} failed to configure on condition: {condition}\n"
+                )
+
+        raise NNCPSuccessStatusNotMet(f"{self.name}. None of the conditions were met.")
```
I do not understand the need for this, can you please explain?
What is the added value to having this vs not having it (and returning None).

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,159 @@
+        ]
+
+        for condition in conditions:
+            if condition["reason"] == self.Conditions.Reason.SUCCESSFULLY_CONFIGURED:
```
This looks wrong.
I think we have already established that the reason is not the criteria to check, it is just the reasoning for why the condition has a specific status. The criteria for success is the relevant condition type and its status (TRUE).

For failures, you could filter by reason, but it is also not enough. You need to assure the type, the status (FALSE) and only then the reason.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,159 @@
+    @retry(
+        wait_timeout=WAIT_FOR_STATUS_TIMEOUT, sleep=WAIT_FOR_STATUS_SLEEP, exceptions_dict={NNCPSuccessStatusNotMet: []}
+    )
+    def wait_for_status_success(self) -> list[dict]:
```
At the moment you are not using the returned value anywhere.
Is there a reason to return a list of dicts?

I do not even understand how this is passing the typing linter. I do not see anything returning a list of dicts in the body.

EdDev replied:
> I'm returning the condition, which is a list of dict.

AFAIK, a condition is a dict, not a list of anything.
The status conditions is a list of dicts.
But now it is gone, so it no longer matters.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,159 @@
+        raise NNCPSuccessStatusNotMet(f"{self.name}. None of the conditions were met.")
+
+    @staticmethod
+    def dict_normalization(data: list[tuple[str, Any]]) -> dict[str, Any]:
```
This should be "private".

Anatw replied:
Ack. BTW, why isn't protected enough?

EdDev replied:
I meant one underscore and not name mangling.
Please set it to one underscore.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,159 @@
+                "nodeSelector": self._node_selector if self._node_selector else {},
+            })
+
+    def _delete_interfaces(self) -> None:
```
Please order the methods for top to bottom such that the caller is before the callee.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,159 @@
+            })
+
+    def _delete_interfaces(self) -> None:
+        desired_state = {}
```
I do not think you need this, you can fill it up directly. But if there is a problem to do it, just do this very close to the usage.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,159 @@
+
+    def _delete_interfaces(self) -> None:
+        desired_state = {}
+        for iface in self._desired_state.interfaces:
+            iface.state = self.Interface.State.ABSENT
```
This is overriding the desired state, which may be a problem.
You better do a deepcopy here.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,159 @@
+            self._delete_interfaces()
+            self.wait_for_status_success()
+        except Exception as exp:
+            self.logger.error(exp)
```
Is this raising an exception?
If not, why you decided to "absorb" it?
What will happen if this fails to the rest of the tests?

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,159 @@
+    pass
+
+
+class NetLibsNodeNetworkConfigurationPolicy(NodeNetworkConfigurationPolicy):
```
The `NetLibs` prefix is not good and make no sense to me.

I see two options:
- Alias the import:
`from ocp_resources.node_network_configuration_policy_latest import NodeNetworkConfigurationPolicy as nncp`
- Call it `nncp`

EdDev replied:
flake8 is wrong, but no point in arguing with a linter in this case.

Anatw commented:
Change: rebase on main.

Anatw commented:
[Changes](https://github.com/RedHatQE/openshift-virtualization-tests/compare/0f24256cf37265193774acda871ca7daba605bd0..fd8ecf132815420170cd5a111d94d8e62694fbeb):
 - Rename global variables.
 - Change the order of the dataclasses for better readability.
 - Rename NetLibsNodeNetworkConfigurationPolicy to NodeNetworkConfigurationPolicy (and change the import of the wrapper NodeNetworkConfigurationPolicy to nncp).
 - Refactor the wait_for_status_success method to rely on status and type before reason.
 - Inside class NodeNetworkConfigurationPolicy, change the order of functions so the caller is before the callee.
 - In _delete_interfaces remove desired_state initialization and assign directly.
 - In the clean_up, remove the try/catch redundant mechanism.
 - Set dict_normalization as a private method.
 - Add a deepcopy operation to the interfaces in _delete_interfaces.

polarion-jenkins reviewed (approved):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/7538

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+        """Filter out none values and converts key characters containing underscores into dashes."""
+        return {key.replace("_", "-"): val for (key, val) in data if val is not None}
+
+    def clean_up(self) -> bool:
+        self._delete_interfaces()
+        self.wait_for_status_success()
+        return super().clean_up()
```
@myakove , how do you treat a teardown which may fail?
In this case, the teardown includes an intermediate step, but the question still stands for any other teardown (because the object deletion may timeout or be rejected).

I am guessing this is super rare and it will be noticeable due to the raised exception. However, I do not see what can be done here except maybe a retry or a stop to the test-run.

EdDev replied:
This has been discussed offline: The lib object should let the exception raise and it will be up to the caller to do something special.

In general, there is nothing that can be done with this.
If it fails, we may leave leftovers.
To mitigate this, in a followup, we should randomize the names of the interfaces and the nncp itself, attempting to avoid a future collision even if something bad happened.
Said that, failure on edit or teardown is not seen on execution, so the priority to handle it is not high.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+            })
+
+    @staticmethod
+    def __dict_normalization(data: list[tuple[str, Any]]) -> dict[str, Any]:
```
Ref for visibility: https://github.com/RedHatQE/openshift-virtualization-tests/pull/218#discussion_r1974933672

sbahar619 replied:
You can remove the double underscores (__) since they aren’t necessary here, and will prevent name mangling. A single underscore (_) will do the job for internal/private use.

EdDev commented on tests/network/cluster_addons_operator/test_network_addons_operator.py:
```diff
@@ -149,21 +159,34 @@ def check_components(network_addons_config_scope_session):
+        name="test-network-operator",
+        desired_state=desired_state,
         node_selector=get_node_selector_dict(node_selector=worker_node1.hostname),
     ) as br_dev:
```
Am I missing something or this is an `nncp` and not a bridge device?

EdDev commented on tests/network/cluster_addons_operator/test_network_addons_operator.py:
```diff
@@ -149,21 +159,34 @@ def check_components(network_addons_config_scope_session):
 
 @pytest.fixture(scope="module")
 def net_add_op_br1test_nad(namespace, net_add_op_bridge_device):
+    interface_name = net_add_op_bridge_device.instance["spec"]["desiredState"]["interfaces"][0]["name"]
```
1. You do not need to re-read the nncp object. It is returned after it has been successfully deployed.
2. You should use a getter (as a property) to fetch the spec (`desired_state`) and access the relevant field.

E.g.
```python
interface_name = nncp_with_bridge.desired_state.interfaces[0].name
```

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+            name=name,
+        )
+        self._desired_state = desired_state
+        self._node_selector = node_selector
```
You can remove all the logic behind node_selector and just pass it to the base class, it is already being handled there.

EdDev replied:
To be more precise, `to_dict` is not needed, you can pass the processed data to the parent.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+
+    def clean_up(self) -> bool:
+        self._delete_interfaces()
+        self.wait_for_status_success()
```
As I understand it correctly, if we end up with failure here, the object deletion is not being executed, is that intentional?

EdDev replied:
Please look at the previous review rounds and the thread below that discusses this.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+            if condition["type"] == Resource.Condition.AVAILABLE:
+                self.logger.info(f"{self.kind} {self.name} configured successfully")
+                return True
+            if (condition["type"] == Nncp.Condition.DEGRADED or condition["type"] == Nncp.Condition.FAILING) and (
+                condition["reason"] == self.Conditions.Reason.NO_MATCHING_NODE
+                or condition["reason"] == self.Conditions.Reason.FAILED_TO_CONFIGURE
+            ):
```
```suggestion
            if (condition["type"] in {Nncp.Condition.DEGRADED, Nncp.Condition.FAILING}) and (
                            condition["reason"]
                            in {
                                self.Conditions.Reason.NO_MATCHING_NODE,
                                self.Conditions.Reason.FAILED_TO_CONFIGURE,
                            }
                        ):
```

qinqon commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+                self.logger.info(f"{self.kind} {self.name} configured successfully")
+                return True
+            if (condition["type"] == Nncp.Condition.DEGRADED or condition["type"] == Nncp.Condition.FAILING) and (
+                condition["reason"] == self.Conditions.Reason.NO_MATCHING_NODE
+                or condition["reason"] == self.Conditions.Reason.FAILED_TO_CONFIGURE
```
You should check just Type, the Reason and Message are kind of informative and should "just" be shown as part of the error in the tests, from a user perspective the give them hints about what to fix, for example if NO_MATCHING_NODE maybe they use a bad nodeSelector or they deployed kubernetes-nmstate wrongly.

EdDev replied:
Thank you!

qinqon replied:
> @qinqon , once we determine the condition is of type `degraded` with status `true`, is there a need to check for the reason? Degraded means it is a stable state of failure, right?

The kubernetes-nmstate implements rollback mechanism so network goes back to previews config, meaning that cluster should be stable network wise, if this is not happening then is a bug for kubernetes-nmstate or nmstate.

Anatw commented:
Change: rebase on main.

polarion-jenkins reviewed (approved):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/7600

Anatw commented:
[Changes](https://github.com/RedHatQE/openshift-virtualization-tests/compare/036b366d821056739d25d4fd26e098be1e3223d8..2d459d5affbd5b60296f9862c2b283c14d714f48):
 - Move API link into the DesiredState class docstring.
 - In the DesiredState class, change interfaces to hold an empty list by default.
 - In NodeNetworkConfigurationPolicy class, send desired_state and node_selector to super().__init__().
 - Add a property method interfaces getter.
 - Refactor __dict_normalization name to a single underscore.
 - Change the order of methods in the NodeNetworkConfigurationPolicy class for better readability.
 - In wait_for_status_success, remove any reference to reasons, we should only rely on the conditions' status and type.
 - In net_add_op_bridge_device, rename br_dev to nncp.
 - In net_add_op_br1test_nad, use the interfaces getter.

polarion-jenkins reviewed (approved):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/7623

Anatw commented:
Change: rebase on main.

Anatw commented:
Change: fix pre-commit (add return value to wait_for_status_success).

polarion-jenkins reviewed (dismissed):
D/S test `tox -e verify-bugs-are-open` passed: cnv-tests-tox-executor/7651

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+@dataclass
+class Bridge:
+    options: BridgeOptions | None = None
+    port: list[Port] | None = None
```
Make it plural:
ports

EdDev replied:
The spec [1] defines it like this.
It is singular.

[1] https://nmstate.io/examples.html#interface-linux-bridge

sbahar619 replied:
Right, thanks, please ignore my comment then.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+@dataclass
+class Bridge:
+    options: BridgeOptions | None = None
+    port: list[Port] | None = None
```
To avoid shared mutable state, consider using field(default_factory=list) for the port attribute.

EdDev replied:
No, it is not a required field and should not take any such default.
The default is not to specify it and therefore it should stay `None`.

sbahar619 replied:
Right, thanks🙏🏼

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+    @retry(
+        wait_timeout=WAIT_FOR_STATUS_TIMEOUT_IN_SEC,
+        sleep=WAIT_FOR_STATUS_INTERVAL_IN_SEC,
+        exceptions_dict={NNCPSuccessStatusNotMet: []},
```
NNCPSuccessStatusNotMet is redundant, it is not being raised anywhere.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+    interfaces: list[Interface] = field(default_factory=list)
+
+
+class NNCPSuccessStatusNotMet(Exception):
```
Not really being used.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+    NodeNetworkConfigurationPolicy object.
+    """
+
+    class Conditions:
```
This whole class is not used.

sbahar619 commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+            node_selector (dict, optional): A node selector that specifies the nodes to apply the node network
+                configuration policy to.
+        """
+        super().__init__(name=name, desired_state=desired_state, node_selector=node_selector)
```
No need to pass desired_state to the base class, as it is already handled in the child class.

EdDev replied:
This implementation is wrong, I will work on this with @Anatw offline.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+from ocp_resources.resource import Resource, ResourceEditor
+from timeout_sampler import retry
+
+WAIT_FOR_STATUS_TIMEOUT_IN_SEC = 90
```
Classification: nit

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+        self,
+        name: str,
+        desired_state: DesiredState,
+        node_selector: dict[str, str] | None = None,
```
Consistency?
Lets start with correctness.
This seems correct to me and is more accurate.

EdDev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,153 @@
+                self.logger.info(f"{self.kind} {self.name} configured successfully")
+                return True
+            if condition["type"] == Nncp.Condition.DEGRADED:
+                raise NNCPConfigurationFailed(f"{self.kind} {self.name} failed to configure on condition:\n{condition}")
```
Why do we need the `kind` part? The fact it is an NNCP is embedded in the exception class. Most of the test is redundant really.

Anatw commented:
[Changes](https://github.com/RedHatQE/openshift-virtualization-tests/compare/42085fb54ea15de9dd77d89f4cd63fa6aaa6e4c2..36bef94d6e3f354e5117ab01492fd993b4dbe2fa):
 - Rename global variables.
 - In dataclasses - set all non-requested values to default None.
 - Refactor DesiredState docstring.
 - Remove NNCPSuccessStatusNotMet.
 - In NodeNetworkConfigurationPolicy, remove class Conditions.
 - In NodeNetworkConfigurationPolicy, refactor init to send to super().init() the desired_data as dict and not as dataclass.
 - Remove the redundant to_dict method.
 - Remove the redundant interfaces method.
 - Add a desired_state_spec method.
 - In wait_for_status_success, replace the exceptions_dict exception to a general exception, only there so that any other error will be raised.
 - Replace self.Condition with Nncp.Condition.
 - Refactor logging in NodeNetworkConfigurationPolicy.
 - Remove duplication in _delete_interfaces.
 - Refactor test_network_addons_operator.py according to the changes.

EdDev reviewed (approved):
Thank you!

yossisegev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,140 @@
+        )
+
+    @staticmethod
+    def _dict_normalization(data: list[tuple[str, Any]]) -> dict[str, Any]:
```
I'm not sure how I feel about this method... ;)
The caller is responsible for calling this class with valid keys and values, and if they don't do that - the attempt to create the resource should fail with the appropriate error (which kubernetes already does).
I don't think the resource class should cover-up for the caller sending invalid parameters.

EdDev replied:
I do not understand your comment.
This method is used to drop `None` values and convert `_` field names to `-` in order to fit the use of `dataclass`.
basically, it is part of the mechanism to convert a structured data (`dataclass`) to a dict that is used to call the API.

It has nothing to do with the user calling anything. The user is building the structured data and if that is wrong the API will be the one who will fail it.

Anatw replied:
https://github.com/RedHatQE/openshift-virtualization-tests/pull/218#discussion_r1983385143
Marking this conversation as resolved.

yossisegev commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,140 @@
+    def _delete_interfaces(self) -> None:
+        desired_state = deepcopy(self.desired_state)
+        for iface in desired_state["interfaces"]:
+            iface["state"] = Nncp.Interface.State.ABSENT
```
Maybe we should limit that only for specific interface types.
If the original NNCP is supposed to modify something in a node's interface, e.g. setting `eno3` to `down`, we wouldn't want this scenario to end with an attempt to delete `eno3`.
And more generally - we don't want to try deleting any interface that has always been there.

EdDev replied:
You are right, but this is following the current logic and going beyond that is a non-trivial follow-up.

In general, what you suggest here is possible.
Do we need it? Only time will tell (or you will tell us :) ).

yossisegev replied:
After discussing it offline with @EdDev - I can live with this change as it is presented above for 2 reasons:
- It actually preserves the same logic we used until now, which haven't resulted with any attempt to delete any day-0 interface.
- Checking the interface type might not be enough, for example - the default `br-ex` is of type `ovs-bridge`, and we shouldn't attempt to remove it.

Anatw replied:
Marking this conversation as resolved.

sbahar619 reviewed (approved):
Thanks!!

EdDev commented:
@dbasunag , @myakove , this PR is ready.

I have discussed the topic of using objects that enhance the wrapper resources with @myakove .
We agreed that it is a good practice to innovate and evolve such abstractions on the tests repo, utilizing it for a period of time and wait for its adoption & stabilization before promoting it to the wrapper (which should be the end goal).

Therefore, I think this change is good to betaken into the repo, with a goal of gradually replacing the existing legacy NNCP object and once it reaches maturity, to promote it to the wrapper.

As with all the rest of the helpers under `/libs/net`, it will be owned by the sig-network.

Anatw commented:
/verified
iib: 927077
jenkins job: #1879

myakove commented:
> @dbasunag , @myakove , this PR is ready.
> 
> I have discussed the topic of using objects that enhance the wrapper resources with @myakove . We agreed that it is a good practice to innovate and evolve such abstractions on the tests repo, utilizing it for a period of time and wait for its adoption & stabilization before promoting it to the wrapper (which should be the end goal).
> 
> Therefore, I think this change is good to betaken into the repo, with a goal of gradually replacing the existing legacy NNCP object and once it reaches maturity, to promote it to the wrapper.
> 
> As with all the rest of the helpers under `/libs/net`, it will be owned by the sig-network.

Agreed, NNCP is problematic resource, make sure it's working as expected before pushing it to the wrapper is good idea.
About `/libs/net`  will be owned by the sig-network, I like it and I think the rest of the team should follow it.

Anatw commented on libs/net/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,140 @@
+from copy import deepcopy
```
Sure, great new that we'll be able to move forward with this feature testing. Should it be a standalone? I think it better go into a new net-infra folder.

Anatw commented:
Change: rebase on main

Anatw commented:
/build-and-push-container

yossisegev commented on tests/network/libs/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,140 @@
+    def clean_up(self) -> bool:
+        self._delete_interfaces()
+        self.wait_for_status_success()
+        return super().clean_up()
```
Possible code duplication: This is calling the wrapper's `clean_up`, which is also going in a flow of deleting interfaces by setting their status to `ABSENT`.

dbasunag commented on tests/network/libs/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,140 @@
+from copy import deepcopy
```
If network team wants to use this code, then this needs to stay under tests/network. If you decide to modify anything in utility, then at that point, you would need to go by maintainers recommendations at that point. This move is to help with getting this PR merged, as network team has been saying they are blocked without this PR.

Anatw replied:
This code in only for network team usage. Marking this comment as resolved.

Anatw commented on tests/network/cluster_addons_operator/test_network_addons_operator.py:
```diff
@@ -17,6 +17,16 @@
 
```
We should show the new code used in a test - otherwise it's not seen as needed.
For this reason, they must stay together.

yossisegev replied:
Accepted, please mark as resolved.

Anatw commented on tests/network/libs/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,140 @@
+
+    def _delete_interfaces(self) -> None:
+        desired_state = deepcopy(self.desired_state)
+        for iface in desired_state["interfaces"]:
```
This was already discussed. Here https://github.com/RedHatQE/openshift-virtualization-tests/pull/218#discussion_r1981402126

EdDev replied:
You raised it here some time ago: https://github.com/RedHatQE/openshift-virtualization-tests/pull/218#discussion_r1981402126

We agreed to take care of this once we will touch these interfaces as part of the iterative approach.

yossisegev replied:
Very well, accepted.
Kudos for you guys for remembering my past comments :) .
Please resolve this conversation.

EdDev commented on tests/network/cluster_addons_operator/test_network_addons_operator.py:
```diff
@@ -149,21 +159,34 @@ def check_components(network_addons_config_scope_session):
 
 @pytest.fixture(scope="module")
 def net_add_op_br1test_nad(namespace, net_add_op_bridge_device):
+    bridge_name = net_add_op_bridge_device.desired_state_spec.interfaces[0].name
```
The input spec for the NNCP is accessed, something defined at line 163 above.
If we there will be a need for something more generic or if the definition will be far away from this access, then we will need to make it nicer, I agree. But for now, it seems good enough.

yossisegev replied:
I strongly disagree, implementing this way is wrong.
And let's be honest - if we don't change it now, we will never change it, until it causes errors and failures. When this happens, it might take us some time to debug and find the root cause.

EdDev replied:
Ohh and knowing the bridge name in advance as a const is not only ugly, but probably wrong. It will not allow randomizing it as it should be.

And looking up for an interface of a specific type is not much better. Assuming you want to protect from a generic nncp spec (why?), then it will not work for cases of multiple such bridges.

EdDev replied:
I’m fine with the `next` version, just drop the `None`.

Anatw replied:
Changed. marking this thread as resolved.

Anatw commented:
[Change](https://github.com/RedHatQE/openshift-virtualization-tests/compare/80648add3c46f3108e4c57163d5a7680b38f88cd..8b706c4fc3201e4c704464f5a61d7deebda1cb7f): address review comments (change getting the bridge name in the NAD from static to dynamic search).

Anatw commented:
/build-and-push-container

cnv-qe-bot1 commented:
New container for quay.io/openshift-cnv/openshift-virtualization-tests:pr-218 published

Anatw commented:
/verified

Jenkins job: openshift-virtualization-tests-runner/2072
iib: 947266
cluster version: 4.19.0-ec.4

Anatw commented on tests/network/libs/nodenetworkconfigurationpolicy.py:
```diff
@@ -0,0 +1,140 @@
+        return self._desired_state
+
+    def clean_up(self) -> bool:
+        self._delete_interfaces()
```
ACK. I'll address this in a follow up PR.

rnetser replied:
you should call `self.wait_for_configuration_conditions_unknown_or_progressing()` and not use timestamps.
was https://github.com/RedHatQE/openshift-python-wrapper/pull/2372 verified?

> ACK. I'll address this in a follow up PR.
this is new code that is being added, the comment should be addressed in this PR

EdDev replied:
@rnetser , this is diverging to another big topic.
The timestamp is unavoidable because the behavior that was detected is that the condition read after the change is the one from before the change. The controller which is responsible to update it did not refresh its value. Therefore, the timestamp was used to confirm that the value is a fresh one.

Beyond that, the attempt to exist early on stable failure is already in place (at `wait_for_status_success`).
If we can discuss it in an immediate follow, it will be better.

vsibirsk commented:
/unhold

vsibirsk commented:
/hold cancel
